scanpy/external/tl/_sam.py

```diff
@@ -0,0 +1,216 @@
+
+    standardization - str, optional, default 'Normalizer'
+        If 'Normalizer', use sklearn.preprocessing.Normalizer, which
+        normalizes expression data prior to PCA such that each cell has
+        unit L2 norm. If 'StandardScaler', use
+        sklearn.preprocessing.StandardScaler, which normalizes expression
+        data prior to PCA such that each gene has zero mean and unit
+        variance. Otherwise, do not normalize the expression data. We
+        recommend using 'StandardScaler' for large datasets with many
+        expected cell types and 'Normalizer' otherwise.
+
+    num_norm_avg - int, optional, default 50
+        The top 'num_norm_avg' dispersions are averaged to determine the
+        normalization factor when calculating the weights. This prevents
+        genes with large spatial dispersions from skewing the distribution
+        of weights.
+
+    weight_pcs - bool, optional, default True
+        If True, scale the principal components by their eigenvalues. In
+        datasets with many expected cell types, setting this to False might
+        improve the resolution as these cell types might be encoded by low-
+        variance principal components.
+
+    npcs - int, optional, default None,
+        Determines the number of top principal components selected at each
+        iteration of the SAM algorithm. If None, this number is chosen
+        automatically based on the size of the dataset. If weight_pcs is
+        set to True, this parameter primarily affects the runtime of the SAM
+        algorithm (more PCs = longer runtime).
+
+    n_genes - int, optional, default None:
+        Determines the number of top SAM-weighted genes to use at each iteration
+        of the SAM algorithm. If None, this number is chosen automatically
+        based on the size of the dataset. This parameter primarily affects
+        the runtime of the SAM algorithm (more genes = longer runtime).
+    
+    inplace - bool, optional, default True:
+        Set fields in `adata` if True. Otherwise, returns a copy.
+    
+    verbose - bool, optional, default True:
+        If True, displays SAM log statements.
+
+    Returns
+    -------
+    sam - SAM
+        The SAM object
+
+    adata - AnnData
+        `.var['weights']`
+            SAM weights for each gene.
+        `.var['spatial_dispersions']`
+            Spatial dispersions for each gene (these are used to compute the
+            SAM weights)
+        `.var['mask_genes']`
+            If preprocessed with SAM, this boolean vector indicates which genes
+            were filtered out (=False).
+        `.uns['preprocess_args']`
+            Dictionary of parameters used for preprocessing.
+        `.uns['run_args']`
+            Dictionary of parameters used for running SAM.
+        `.uns['pca_obj']`
+            The sklearn.decomposition.PCA object.
+        `.uns['X_processed']`
+            The standardized and SAM-weighted data fed into PCA.
```

Comment:
Why `.uns` and not `obsm` or `layers`?

Reply:
Good catch! I had since changed this to be stored in the SAM object itself as opposed to the AnnData. It's a large, dense array and ended up not wanting to make AnnData too bloated